### PR TITLE
test(icrc_ledger): FI-1368: Upgrade test for ICRC-1 ledgers using golden state

### PR DESCRIPTION
**Pull request by mbjorkqvist**
Add upgrade test for the ICRC-1 ledgers using golden canister state.

### CODE REVIEW

**Comment by mbjorkqvist** [comment]
Earlier GitLab MR here: https://gitlab.com/dfinity-lab/public/ic/-/merge_requests/20093

**Review by maciejdfinity — APPROVED** [review]
I assume this is the exact copy of https://gitlab.com/dfinity-lab/public/ic/-/merge_requests/20093

**Comment by max-dfinity** [comment]
Question about this - is it using a different state backup than the one used for canister upgrade qualification for the rest of the canisters on the subnet?

File: rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        SONIC,
+        TRAX,
+        WATERNEURON,
+        YUKU,
+    ];
+
+    let state_machine = new_state_machine_with_golden_sns_state_or_panic();
+
+    for canister in canisters {
+        upgrade_canister(&state_machine, canister, ledger_wasm.clone());
```
Comment (max-dfinity) [inline rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs:109]: nit, and no change required, but "canister" usually means "canister_id" or a struct representing a canister, not a tuple.

File: rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+fn upgrade_canister(
+    state_machine: &StateMachine,
+    (canister_id_str, canister_name): (&str, &str),
+    ledger_wasm: Vec<u8>,
+) {
+    let canister_id =
+        CanisterId::unchecked_from_principal(PrincipalId::from_str(canister_id_str).unwrap());
+    upgrade_ledger(state_machine, ledger_wasm, canister_id);
```
Comment (max-dfinity) [inline rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs:120]: So we have been doing 2x upgrades in our tests to ensure that the first upgrade does not end up in a state where a second upgrade is not possible (i.e. pre_upgrade hook failure)

I don't know if that's covered elsewhere (haven't finished my review)

**Review by max-dfinity — APPROVED** [review]

File: rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs (hunk 2 above)
Comment (max-dfinity) [reply rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs:120]: I didn't find a place where this is done.  There should be examples in our tests, but we modify the wasm metadata so we can get a different hash, and then we do a second upgrade

**Comment by max-dfinity** [comment]
NNS changes lgtm

File: rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs (hunk 2 above)
Comment (mbjorkqvist) [reply rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs:120]: You're probably referring to [`bump_gzip_timestamp`](https://sourcegraph.com/github.com/dfinity/ic@fe231b38582fd12bb3e238e4e689fafa59b2b14f/-/blob/rs/nns/test_utils/src/itest_helpers.rs?L728-739). I'd heard about it, but I hadn't realized before that a 2x upgrade would help to catch a `pre_upgrade` hook failure - that's a great suggestion, and I'll add it to this test, and consider adding it to our other upgrade tests also!

Comment (mbjorkqvist) [reply rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs:144]: I added a second upgrade, but that second upgrade was then returning a `CanisterInstallCodeRateLimited` error. I asked about this [on Slack](https://dfinity.slack.com/archives/CGZJ7G1J6/p1721300686300469?thread_ts=1717412583.592399&cid=CGZJ7G1J6), and implemented the solution of disabling `rate_limiting_of_instructions`.

Comment (max-dfinity) [reply rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs:144]: Great, glad it was useful!

**Comment by mbjorkqvist** [comment]
Thanks for the reviews and feedback, @maciejdfinity and @max-dfinity!

File: rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs (hunk 1 above)
Comment (mbjorkqvist) [reply rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs]: Since I was anyway making changes I renamed these also, thanks for the suggestion!

**Pull request merged**